File: includes/water_level_receiver.py
import mysql.connector
from datetime import datetime

# First verify serial import works
try:
    import serial
    print("PySerial version:", serial.__version__)
    print("Serial attributes:", dir(serial))  # Should show 'Serial'
except Exception as e:
    print("Serial import error:", e)
    exit(1)

def insert_into_mysql(water_level_data):
    try:
        connection = mysql.connector.connect(
            host='localhost',
            user='root',
            password='',
            database='fews'
        )
        
        cursor = connection.cursor()
        current_time = datetime.now()
        
        query = """INSERT INTO sensor_readings 
                   (water_level, reading_time) 
                   VALUES (%s, %s)"""
        
        cursor.execute(query, (water_level_data, current_time))
        connection.commit()
        print("Data inserted successfully")
        
    except Exception as e:
        print(f"Database error: {e}")
    finally:
        if 'connection' in locals() and connection.is_connected():
            cursor.close()
            connection.close()

try:
    ser = serial.Serial('COM3', 115200, timeout=1)
    print(f"Serial port {ser.name} opened successfully")
    
    try:
        print("Listening for serial data...")
        while True:
            if ser.in_waiting > 0:
                data = ser.readline().decode('utf-8').strip()
                print(f"{data}")
                insert_into_mysql(data)
                
    except KeyboardInterrupt:
        print("\nProgram terminated by user")
        
finally:
    if 'ser' in locals():
        ser.close()
        print("Serial port closed")

File: includes/water_level_receiver.php
<?php
header("Access-Control-Allow-Origin: *");
header("Content-Type: application/json");

// Get raw POST data
$content = file_get_contents("php://input");
parse_str($content, $data);

// Validate input - less strict validation since column is VARCHAR
if (!isset($data['water_level'])) {
    http_response_code(400);
    echo json_encode([
        "status" => "error",
        "message" => "Missing water_level parameter"
    ]);
    exit;
}

$servername = "localhost";
$username = "root";
$password = "";
$dbname = "fews";

try {
    $conn = new PDO("mysql:host=$servername;dbname=$dbname", $username, $password);
    $conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

    // Trim and sanitize the input
    $water_level = substr(trim($data['water_level']), 0, 100); // Limit to reasonable length
    
    $stmt = $conn->prepare("INSERT INTO sensor_readings (water_level, reading_time) VALUES (:water_level, NOW())");
    $stmt->bindParam(':water_level', $water_level, PDO::PARAM_STR);
    $stmt->execute();

    echo json_encode([
        "status" => "success",
        "message" => "Data inserted successfully",
        "data" => [
            "water_level" => $water_level,
            "reading_time" => date('Y-m-d H:i:s')
        ]
    ]);
} catch(PDOException $e) {
    http_response_code(500);
    echo json_encode([
        "status" => "error",
        "message" => "Database error: " . $e->getMessage()
    ]);
}
?>